iroh-net/src/magicsock.rs

```diff
@@ -278,6 +286,664 @@ impl Inner {
     fn public_key(&self) -> PublicKey {
         self.secret_key.public()
     }
+
+    /// Get the cached version of the Ipv4 and Ipv6 addrs of the current connection.
+    fn local_addr(&self) -> (SocketAddr, Option<SocketAddr>) {
+        *self.local_addrs.read().unwrap()
+    }
+    fn normalized_local_addr(&self) -> io::Result<SocketAddr> {
+        let (v4, v6) = self.local_addr();
+        let addr = if let Some(v6) = v6 { v6 } else { v4 };
+        Ok(addr)
+    }
+
+    #[instrument(skip_all, fields(me = %self.me))]
+    fn poll_send(
+        &self,
+        cx: &mut Context,
+        transmits: &[quinn_udp::Transmit],
+    ) -> Poll<io::Result<usize>> {
+        let bytes_total: usize = transmits.iter().map(|t| t.contents.len()).sum();
+        inc_by!(MagicsockMetrics, send_data, bytes_total as _);
+
+        if self.is_closed() {
+            inc_by!(MagicsockMetrics, send_data_network_down, bytes_total as _);
+            return Poll::Ready(Err(io::Error::new(
+                io::ErrorKind::NotConnected,
+                "connection closed",
+            )));
+        }
+
+        let mut n = 0;
+        if transmits.is_empty() {
+            return Poll::Ready(Ok(n));
+        }
+
+        trace!(
+            "sending:\n{}",
+            transmits.iter().fold(
+                String::with_capacity(transmits.len() * 50),
+                |mut final_repr, t| {
+                    final_repr.push_str(
+                        format!(
+                            "  dest: {}, src: {:?}, content_len: {}\n",
+                            QuicMappedAddr(t.destination),
+                            t.src_ip,
+                            t.contents.len()
+                        )
+                        .as_str(),
+                    );
+                    final_repr
+                }
+            )
+        );
+
+        let dest = transmits[0].destination;
+        for transmit in transmits.iter() {
+            if transmit.destination != dest {
+                break;
+            }
+            n += 1;
+        }
+        // TODO: This is the remaining alloc on the hot path for send.
+        // Unfortunately I don't see a way around this because we have do modify the transmits.
+        let mut transmits = transmits[..n].to_vec();
```

Comment:
we could use a fixed buffer, that might help

Reply:
Added a fixed buffer, so one alloc less. Instead there's a `parking_lot::Mutex::lock` but it should have no contention ever so fast.

We will still allocate when sending over Derp, because the derp actor expectes a `Vec<Bytes>`.  Not sure if there's a way to reduce allocs here as well, likely not - would defer to followup.

Reply:
If this is always or frequently a single Bytes we could reach for mallvec for such issues.